src/expint.jl

```diff
@@ -0,0 +1,331 @@
+        
+        if abs(asum_prev - asum) < 1e-15
+            break
+        end
+        
+        Δ_prod_fact *= -Δ / (k+2)
+        
+        iters += 1
+        k += 1
+    end
+
+    res = exp(-z₀) * asum
+    return res
+end
+
+# series about origin
+# https://functions.wolfram.com/GammaBetaErf/ExpIntegralE/06/01/04/01/01/0003/
+function En_expand_origin(x::Number, ν::Number)
+    if isreal(ν) && floor(ν) == ν
+        # go to special case for integer ν
+        return En_expand_origin(x, Int(ν))
+    end
+    gammaterm = gamma(1 - ν)
+    if abs(gammaterm) != 0
+        # don't multiply by NaN
+        gammaterm *= x^(ν-1)
+    end
+    frac = 1
+    sumterm = frac / (1 - ν)
+    k, maxiter = 1, 100
+    while k < maxiter
+        frac *= -x / k
+        prev = sumterm
+        sumterm += frac / (k + 1 - ν)
+        if abs(sumterm - prev) < 1e-15 * abs(prev)
+            break
+        end
+        k += 1
+    end
+    
+    return gammaterm - sumterm
+end
+
+# series about the origin, special case for integer n
+function En_expand_origin(x::Number, n::Integer)
+    gammaterm = 1
+    for i = 1:n-1
+        gammaterm *= -x / i
+    end
+
+    gammaterm *= digamma(n) - log(x)
+    sumterm = n == 1 ? 0 : 1 / (1 - n)
+    frac = 1
+    k, maxiter = 1, 100
+    while k < maxiter
+        frac *= -x / k
+        # skip term with zero denominator
+        if k != n-1
+            prev = sumterm
+            sumterm += frac / (k + 1 - n)
+            if abs(sumterm - prev) < 1e-15 * abs(prev)
+                break
+            end
+        end
+        k += 1
+    end
+    return gammaterm - sumterm
+end
+
+# can find imaginary part of E_ν(x) for x on negative real axis analytically
+function imagbranchcut(x::Number, ν::Number)
+    a = real(x)
+    impart = π * im * exp(-π*im*ν) * a^(ν-1) / gamma(ν)
+    # exp(n*log(z) - loggamma(n))
+    return imag(impart) * im # get rid of any real error
+end
+
+const ORIGIN_EXPAND_THRESH = 3
+function En(ν::Number, x::Number, niter::Int=1000, debug=false)
+    if x == 0.0
+        if real(ν) > 0
+            return 1.0 / (ν - 1)
+        else
+            return Inf
+        end
+    end
+    if ν == 0
+        return exp(-x) / x
+    end
+    # asymptotic test
+    if exp(-x) / x == 0
+        return zero(x)
+    end
+
+    if abs(x) < ORIGIN_EXPAND_THRESH
+        return En_expand_origin(x, ν)
+    end
+    E_guess, _, g = En_cf(x, ν, niter)
+    if g
+        return E_guess
+    end
+    if real(x) > 0
+        res, i, _ = En_cf(x, ν, niter)
+        return res
+    elseif real(x) < 0
+        doconj = imag(x) < 0
+        rex, imx = real(x), abs(imag(x))
+        x = doconj ? conj(x) : x
+        ν = doconj ? conj(ν) : ν
+        
+        # empirical boundary for 500 iterations
+        boundary = min(1 + 0.5*abs(ν), 50)
+        if imx > boundary
+            res, i, _ = En_cf(x, ν, niter)
+            return doconj ? conj(res) : res
+        else
+            # iterate with taylor
```

Comment:
Some comment here that this is similar to the algorithm suggested by Amos, and give a reference?

In general, some references might be helpful, as well as more comments in the code about why you are using one approximation vs. another.